lib/Sema/CSSimplify.cpp

```diff
@@ -2793,6 +2793,26 @@ bool ConstraintSystem::repairFailures(
   return !conversionsOrFixes.empty();
 }
 
+static bool
+shouldDiagnoseUnnecessaryExplicitCoercion(ConstraintSystem &CS,
+                                          ConstraintKind kind,
+                                          ConstraintLocatorBuilder locator) {
+  auto locatorPtr = locator.getBaseLocator();
+  if (kind >= ConstraintKind::Conversion) {
+    if (auto last = locator.last()) {
+      if (last->is<LocatorPathElt::ExplicitTypeCoercion>()) {
+        return llvm::none_of(CS.getFixes(),
+                             [&locatorPtr](const ConstraintFix *fix) -> bool {
+          if (fix->getLocator() == locatorPtr)
+            return true;
+          return fix->getLocator()->getAnchor() == locatorPtr->getAnchor();
```

Comment:
Why is this check necessary?

Reply:
There was a case
```swift
var s2 = ns ?? "str" as String as String
```
 where it was triggering twice, so this check was to limit one warning for expression only in this case.

Reply:
I think I found the "issue" why the diagnostic is being emitted twice for this case: 
Actually is for a case there is a failure conversion in the expression like:
```swift
 var s2 = ns ?? "str" as String as String 
Double(1) as Double as String
```
So there is a failure and when trying to salvage the system it ends up here on [diagnosing](https://github.com/apple/swift/blob/4ade7ae11f628b5f33d51de8e90489aa45039a2b/lib/Sema/ConstraintSystem.cpp#L2471) which is leading to a re-type checking later, causing this Fixit path to be called twice with two different constraint systems.

I could only think in 3 options here, which are:
1. Check also for a baseCS in the `attempt` method so if the cs has a base there it is probably on a path of re-type checking. (or some other way to simply check this)
2. Revert and check for the anchor of the fix (like this one) and limit one warn for anchor expr.
3. Emit some kind of check [there](https://github.com/apple/swift/blob/4ade7ae11f628b5f33d51de8e90489aa45039a2b/lib/Sema/ConstraintSystem.cpp#L2471) to avoid going down that path if the expr matches that fixit requirements. Not sure how it would affect other diagnostics.

Obs: Already implemented option 1 in last commit. But let me know if there is a better option :)
cc @xedin